version: '3.8'

x-common:
  environment:
    &common-env
    POSTGRES_DB: ${POSTGRES_DB}
    POSTGRES_USER: ${POSTGRES_USER}
    POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
    API_TOKEN: ${API_TOKEN}
    API_URL: ${API_URL}

services:

  rdbms:
    container_name: rdbms
    image: postgres:16.2-alpine
    restart: on-failure
    ports:
      - '9876:5432'
    environment:
      <<: *common-env
    volumes:
      - ./db:/docker-entrypoint-initdb.d
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -d ${POSTGRES_DB} -U ${POSTGRES_USER}"]
      interval: 30s
      retries: 10

  # nosql:
  #   image: mongo:7.0.7
  #   container_name: mongo
  #   ports:
  #     - "27017:27017"

  # mongo-admin:
  #   image: mongo-express:1.0.2-20-alpine3.19
  #   container_name: mongo-admin
  #   ports:
  #     - "8081:8081"
  #   depends_on:
  #     - nosql

  ingestion:
    container_name: python_runtime
    build:
      context: ./
      dockerfile: python.Dockerfile
    environment:
      <<: *common-env
    volumes:
      - ./scripts:/app
    command: >
      "sleep 5 && python main.py"
    depends_on:
      - rdbms
    tty: true

  dbt:
    container_name: dbt
    profiles:
      - transform
    build:
      context: ./
      dockerfile: dbt.Dockerfile
    volumes:
      - ./dbt:/dbt
      - ./dbt/.dbt:/root/.dbt
    depends_on:
      - rdbms
      - ingestion
    environment:
      <<: *common-env
    working_dir: /dbt/user_events
    command: run
    tty: true